astarte/device/device.py

```diff
@@ -369,6 +372,14 @@ def __on_disconnect(self, client, userdata, rc):
         # If rc was explicit, stop the loop (after the callback)
         if rc == 0:
             self.__mqtt_client.loop_stop()
+        # Else check certificate expiration and try reconnection
+        elif crypto.certificate_is_expired(os.path.join(self.__persistency_dir, self.__device_id,
+                                                        "crypto")):
+            self.disconnect()
```

Comment:
This is useless: in line 362 `__is_connected` is set to `False`, so this call will just return.

What you do have to call instead is `self.__mqtt_client.loop_stop()`: `self_connect()` spawns a new thread with `self.__mqtt_client.loop_start()`, and if you don't stop the old one you're leaking a thread every time you renew the certificate after its expiry.

Reply:
fixed